wgpu-hal/src/dx12/mod.rs

```diff
@@ -44,17 +44,254 @@ mod suballocation;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory1") }?;
+
+        let mut result__ = core::ptr::null_mut();
+        Ok((func)(&Dxgi::IDXGIFactory1::IID, &mut result__)
+            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
+    }
+
+    pub fn create_factory2(
+        &self,
+        factory_flags: Dxgi::DXGI_CREATE_FACTORY_FLAGS,
+    ) -> Result<windows::core::Result<Dxgi::IDXGIFactory4>, libloading::Error> {
+        // Calls windows::Win32::Graphics::Dxgi::CreateDXGIFactory2 on dxgi.dll
+        type Fun = extern "system" fn(
+            flags: Dxgi::DXGI_CREATE_FACTORY_FLAGS,
+            riid: *const windows_core::GUID,
+            ppfactory: *mut *mut core::ffi::c_void,
+        ) -> windows_core::HRESULT;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory2") }?;
+
+        let mut result__ = core::ptr::null_mut();
+        Ok(
+            (func)(factory_flags, &Dxgi::IDXGIFactory4::IID, &mut result__)
+                .and_then(|| unsafe { windows_core::Type::from_abi(result__) }),
+        )
+    }
+
+    pub fn create_factory_media(
+        &self,
+    ) -> Result<windows::core::Result<Dxgi::IDXGIFactoryMedia>, libloading::Error> {
+        // Calls windows::Win32::Graphics::Dxgi::CreateDXGIFactory1 on dxgi.dll
+        type Fun = extern "system" fn(
+            riid: *const windows_core::GUID,
+            ppfactory: *mut *mut core::ffi::c_void,
+        ) -> windows_core::HRESULT;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"CreateDXGIFactory1") }?;
+
+        let mut result__ = core::ptr::null_mut();
+        // https://learn.microsoft.com/en-us/windows/win32/api/dxgi1_3/nn-dxgi1_3-idxgifactorymedia
+        Ok((func)(&Dxgi::IDXGIFactoryMedia::IID, &mut result__)
+            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
+    }
+}
+
+/// Create a temporary "owned" copy inside a [`mem::ManuallyDrop`] without increasing the refcount or
+/// moving away the source variable.
+///
+/// This is a common pattern when needing to pass interface pointers ("borrows") into Windows
+/// structs.  Moving/cloning ownership is impossible/inconvenient because:
+///
+/// - The caller does _not_ assume ownership (and decrement the refcount at a later time);
+/// - Unnecessarily increasing and decrementing the refcount;
+/// - [`Drop`] destructors cannot run inside `union` structures (when the created structure is
+///   implicitly dropped after a call).
+///
+/// See also <https://github.com/microsoft/windows-rs/pull/2361#discussion_r1150799401> and
+/// <https://github.com/microsoft/windows-rs/issues/2386>.
+///
+/// # Safety
+/// Performs a [`mem::transmute_copy()`] on a refcounted [`Interface`] type.  The returned
+/// [`mem::ManuallyDrop`] should _not_ be dropped.
+pub unsafe fn borrow_interface_temporarily<Src: Interface, Dst>(
+    src: &Src,
+) -> mem::ManuallyDrop<Option<Dst>> {
```

Comment:
Probably worthwhile having a `for<'a> &'a Dst: From<&'a Src>` bound on this function to make limit the scale of mistake you can make with it.

Reply:
I think I needed an interface cast in another crate where this was introduced (and should apply the suggestion there - quick reminder that interface hierarchies weren't a thing in `windows-rs` back then) but no such thing is ever needed in `wgpu` it seems so `Dst` is dropped entirely in favour of a renamed `Src`.